### ISSUE TITLE
Failing cases in JSX mode

### ISSUE THREAD

**Issue by dabbott** [codemirror/CodeMirror#3745]
Following up on: https://github.com/codemirror/CodeMirror/pull/3742

Three from the jsx/index.html page:
1. Boolean attributes are not recognized properly (note the red self-closing `/>`)
   <img width="600" alt="screen shot 2015-12-29 at 9 41 02 am" src="https://cloud.githubusercontent.com/assets/1198882/12038976/6d3289a8-ae11-11e5-9c15-516930ddc12a.png">
2. Block comments styled incorrectly (should be the brown comment color)
3. Line comments are not recognized properly (note the red self-closing `/>`)
   <img width="599" alt="screen shot 2015-12-29 at 9 41 30 am" src="https://cloud.githubusercontent.com/assets/1198882/12038996/8eada342-ae11-11e5-9f8a-7de73b498e83.png">

Two other things I've found:

4 . Elements should be allowed as attributes
<img width="600" alt="screen shot 2015-12-29 at 9 42 54 am" src="https://cloud.githubusercontent.com/assets/1198882/12039046/f650f508-ae11-11e5-9bc9-f7bdebe97911.png">

5 . Indentation in JS expressions should be relative to the indentation level of JSX elements. I believe currently they get indented to prior JS indentation level. (this issue is highest priority, if I had to choose)

Current behavior gif here:
![js-expression-indentation](https://cloud.githubusercontent.com/assets/1198882/12039235/69cd6312-ae13-11e5-9aa8-9717118f3056.gif)

Correct position here:
<img width="213" alt="screen shot 2015-12-29 at 10 05 11 am" src="https://cloud.githubusercontent.com/assets/1198882/12039262/b9f03f9a-ae13-11e5-86fb-264053c93a83.png">

Another example of current behavior (not followed by } char):
![js-indentation-2](https://cloud.githubusercontent.com/assets/1198882/12039320/37f54d54-ae14-11e5-8e80-50b4ab480e42.gif)

Correct position here:
<img width="251" alt="screen shot 2015-12-29 at 10 08 29 am" src="https://cloud.githubusercontent.com/assets/1198882/12039316/3066bfe6-ae14-11e5-8587-36eb0f7ae92c.png">

@parisk @petetnt

**Comment by marijnh** [comment]
I pushed some patches that take care of most of these (though getting indentation right across language boundaries is always tricky, and I'm not sure if the result is what you consider the right way to indent these). But...

>  Elements should be allowed as attributes

What? I saw something like that in the grammar and assumed it was a typo. I guess we're no longer in XML Kansas anymore here, huh? What does this even do? That's going to be somewhat tricky to support since it does not correspond to the XML mode's expectations of what XML looks like.

**Comment by dabbott** [comment]
Awesome, thanks for fixing these so quickly! I'll test and confirm they're fixed. As far as style goes, I'm trusting what facebook shows on the reactjs site as the "proper" way to indent.

Yeah, I dunno, I've never seen anyone write elements as attributes *in any JSX code I've ever seen in the wild over the past year*... I believe it's exactly the same as writing `foo={<Foo />}`, except without the `{}`, for convenience. I don't personally find this case to be valuable or relevant, other than to properly adhere to the spec (which is *not* adhered to by Babel in a few places, as far as I can tell -- although Babel does transpile this case [here](https://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&code=%3CFoo%20bar%3D%3CBar%20%2F%3E%20%2F%3E)).

It's in the [spec](https://facebook.github.io/jsx/) as:
```
JSXAttributeValue :

" JSXDoubleStringCharactersopt "
' JSXSingleStringCharactersopt '
{ AssignmentExpression }
JSXElement
```

IMO, not worth the effort.

**Comment by petetnt** [comment]
I agree that direct `JSXElements` as props (such as `<Foo bar=<Biz /> />`) is a very low priority use case: like @dabbott, I have pretty much never seen them used like that, actually I didn't even know about it until I brainstormed with @lkcampbell about how CodeMirror could handle the props.

**Comment by dabbott** [comment]
1: Boolean attributes :white_check_mark: 

2 & 3: Comments are correct for the list of attributes `<foo /*comment*/>`, but in between an open and close tag they should be colored as regular text instead of comments `<foo>/*not a comment*/</foo>`. Transpiled correctly [here](https://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&code=%3CFoo%3E%0A%20%20%3CFoo%20%2F*comment%20ok*%2F%3E%3C%2FFoo%3E%0A%20%20%3CFoo%3E%2F*should%20not%20be%20a%20comment%20*%2F%3C%2FFoo%3E%0A%20%20%3CFoo%20%0A%20%20%20%20bar%3D%22bar%22%2F%2Fcomment%20ok%0A%20%20%3E%3C%2FFoo%3E%0A%20%20%3CFoo%3E%2F%2Fshould%20not%20be%20a%20comment%20%3C%2FFoo%3E%0A%3C%2FFoo%3E) (though colored wrong).

4: Won't fix, IMO.

5: Much much better. Still a few weird points. I'm not really sure how to systematically describe these other than as random one-offs of how they do/don't behave like I expect... 

Fixed (although, should the closing `}` be electric?)
![js-indentation-3](https://cloud.githubusercontent.com/assets/1198882/12044452/707bf948-ae46-11e5-9501-9d073050a683.gif)

Better, but indents one spot too far, I think.
![js-indentation-4](https://cloud.githubusercontent.com/assets/1198882/12044473/ca3c0cca-ae46-11e5-88cf-8d8237d067d4.gif)

Good (also, closing `}` should be electric?)
![js-indentation-5](https://cloud.githubusercontent.com/assets/1198882/12044481/f1387dea-ae46-11e5-8f71-cb750479ec8f.gif)

Here I expect the closing `}` to align with `showFooter`.
![js-indentation-6](https://cloud.githubusercontent.com/assets/1198882/12044492/1cd4a8f2-ae47-11e5-96e8-65bcdb35203b.gif)

Two more things I found!

6: Attribute indentation doesn't behave as I would expect. I would set `multilineTagIndentPastTag: false` on xml by default when jsx is used. This matches the attribute indentation style Facebook uses.

7: Fat arrow => expression doesn't parse correctly. If surrounded by () it's ok. Fat array => {statements} is also ok.
<img width="458" alt="screen shot 2015-12-29 at 3 17 39 pm" src="https://cloud.githubusercontent.com/assets/1198882/12044443/2175d6ca-ae46-11e5-972f-61d449d3d231.png">

8: Very minor, but something's a little off about the demo page https://codemirror.net/mode/jsx/index.html. The tab character inserts the equivalent of 4 spaces, but adding a newline after a `{` or elsewhere only indents 2 spaces. I would be screenshotting the demo page instead of my own app, but I can't quite tell what's correct and what's not.

**Comment by marijnh** [comment]
> The tab character inserts the equivalent of 4 spaces, but adding a newline after a { or elsewhere only indents 2 spaces.

This is CodeMirror's default behavior. You can configure it with the `indentUnit` and `tabSize` options if you want, but generally you'll want to use shift-tab (auto indent) rather than tab to indent lines.

I fixed comments only being comments inside of tags, and also added support for tag attributes (which, if you see them as a shorthand for `{<tag/>}`, aren't as hard to deal with as I initially thought). I've also disabled `multilineTagIndentPastTag` in the XML submode's configuration.

**Comment by dabbott** [comment]
1: :white_check_mark: Boolean attributes 
2: :white_check_mark: Line comments 
3: :white_check_mark: Block comments 
4: :white_check_mark: `Attr=<Element />`. Great! Glad not to have to restrict what users can do.
5: JS `{}` indentation
6: :white_check_mark: Attr indentation 
7: :white_check_mark: Fat arrow expressions 
8: :white_check_mark: Not an issue, my bad

**Comment by dabbott** [comment]
This one's just a general javascript mode issue I found while testing jsx mode. Object destructuring with a trailing comma was throwing off the rest of the parsing:

<img width="291" alt="screen shot 2015-12-30 at 12 37 44 pm" src="https://cloud.githubusercontent.com/assets/1198882/12057090/5ad5561e-aef2-11e5-98ac-61a957059226.png">

Fixed in #3748:

<img width="291" alt="screen shot 2015-12-30 at 12 41 37 pm" src="https://cloud.githubusercontent.com/assets/1198882/12057130/ba8b536a-aef2-11e5-8931-e571881f0de2.png">

It's [here](http://www.ecma-international.org/ecma-262/6.0/#sec-destructuring-binding-patterns) in the spec. (I thought maybe I was just writing weird code!)

```
13.3.3 Destructuring Binding Patterns

...

ObjectBindingPattern[Yield] :
{ }
{ BindingPropertyList[?Yield] }
{ BindingPropertyList[?Yield] , }
```

**Comment by dabbott** [comment]
I wonder if indentation is off because the open and close brackets are getting parsed in different modes? The `{` is styled as `cm-m-xml`, while the `}` is styled as `cm-m-javascript`. I only noticed this when I tried turning on some addons.

They don't quite match up with the `edit/matchbrackets` addon.
![js-indentation-7](https://cloud.githubusercontent.com/assets/1198882/12061245/dc39ed1c-af35-11e5-90f5-6d41a28db8bf.gif)

They sometimes get automatically closed with `edit/closebrackets` and sometimes don't.
![js-indentation-8](https://cloud.githubusercontent.com/assets/1198882/12061269/757ddb8c-af36-11e5-9cf0-1318d17cb8bc.gif)

**Comment by marijnh** [comment]
> I wonder if indentation is off because the open and close brackets

The brackets are fed to the correct mode, but the way `addModeClass` works, it depends on `innerMode`  returning the correcting thing *before* the token was read. I've updated the mode in 65950ad to address this.

**Comment by dabbott** [comment]
I tweaked the indentation within `{}` for the common cases. Indentation will look like this now:

<img width="463" alt="screen shot 2016-01-06 at 1 52 57 pm" src="https://cloud.githubusercontent.com/assets/1198882/12155924/dda8e382-b47c-11e5-9988-d991dcdf85b7.png">

One thing I wasn't sure how to handle:

<img width="461" alt="screen shot 2016-01-06 at 1 47 34 pm" src="https://cloud.githubusercontent.com/assets/1198882/12155965/09708600-b47d-11e5-8af8-76e66c0419a7.png">

I want to indent differently, depending on whether a `<tag>` is on the same line (and the start of the line) or not. I naively look at the previous token on this line for a `>` ([here](https://github.com/codemirror/CodeMirror/pull/3759/files#diff-8fe55bb9634da263c3fa302cdbefb06cR59)), but that's not quite right (since there can be junk in the way). Is there a better way to do this?

**Comment by marijnh** [comment]
Great. Anything left before we can close this?

**Comment by dabbott** [comment]
One last thing I found: currently the JS within the `{` `}` is getting parsed as a statement, when it should be parsed as an expression. 

E.g. each property in this object is parsed as variable, and it's one long continued statement
<img width="142" alt="screen shot 2016-01-12 at 5 52 53 pm" src="https://cloud.githubusercontent.com/assets/1198882/12282760/8183d5e8-b955-11e5-8844-4c42e6599af6.png">

Should look like this:
<img width="143" alt="screen shot 2016-01-12 at 5 51 54 pm" src="https://cloud.githubusercontent.com/assets/1198882/12282799/d4337b4a-b955-11e5-8e83-c3424e6a7476.png">

I took a stab at fixing by adding a hook
```javascript
expectExpression: function(state) {
    state.cc.push(expressionNoComma)
}
```
to JS mode. Which mostly works, except that the curly braces are getting parsed by the JS mode also. So I tried:

1) Parsing the curly braces manually in JSX mode (and they end up cm-m-xml), so that expecting an expression works correctly. Which does work, but the logic is a little ugly, and I will have to mess with indentation a bit more, since now using the JS `lexical.indented` isn't quite right. https://github.com/codemirror/CodeMirror/compare/master...dabbott:parse-braces-in-jsx?expand=1

or

2) Do it all as a hook to JS mode, but the hook will be very specific to JSX. And I will probably have to pass in `indent` param and assign it to `lexical.indented`, or something, since this breaks indentation in an edge case. https://github.com/codemirror/CodeMirror/compare/master...dabbott:expect-expression-block?expand=1

In both cases I would still need to tweak indentation a bit more. Or maybe theres a better way entirely. What do you think?

**Comment by marijnh** [comment]
I think parsing the braces in the JSX mode is a sane approach (they aren't _really_ part of the JS, anyway). You could set the JS mode to JSON to get it to parse the top-level as an expression, but that'll have some undesirable effect (such as turning off the comment info), so I think there it would make sense to extend the JS mode to also have a configuration that says 'expect an expression', without it being full-on JSON mode.

**Comment by julienw** [comment]
I can add another slightly broken example here:
```javascript
  render() {
    this._scheduleDraw();
    return (
      <div className={this.props.className}>
        <canvas
          className={classNames(
            `${this.props.className}Canvas`,
            'threadStackGraphCanvas'
          )}
          ref={ref => (this._canvas = ref)}
          onMouseUp={this._onMouseUp}
        />
      </div>
    );
  }
```
And the result in current CodeMirror:
![screenshot-2017-10-30 codemirror](https://user-images.githubusercontent.com/454175/32163705-6d9e2b06-bd5d-11e7-9761-386533316ff9.png)

You can see some of the colors are still wrong: see the angle bracket in the `ref`'s arrow function, or `onMouseUp` not having the right color, or the closing `div` tag.

**Comment by marijnh** [comment]
Pasting that code into http://codemirror.net/mode/jsx/ yields different results than what you screenshotted, and actually looks fine.

**Comment by julienw** [comment]
Ah right, I just used Codemirror's home page's demo, so likely was in HTML mode... Sorry about the disturbance...

Is there any drawback using JSX mode for all JS code ?

**Comment by marijnh** [comment]
JSX is a superset, so probably not.

**Comment by julienw** [comment]
Thanks ! Maybe it's not the right location to ask this, but why don't you merge JS and JSX modes, then ?

**Comment by adrianheine** [comment]
What do you mean by merging? If you're talking about code, `mode/javascript/javascript.js` already has 875 lines of code and defines JavaScript, TypeScript, JSON and JSONLD modes. It wouldn't make sense to put the ~150 lines for JSX in that file, too.

**Comment by adrianheine** [comment]
If you select JavaScript mode, you get JavaScript, not TypeScript or JSX. These are different languages, they just happen to share a lot.
